src/rdkafka_broker.c

```diff
@@ -351,33 +348,170 @@ void rd_kafka_broker_set_state (rd_kafka_broker_t *rkb, int state) {
+                                       const char *fmt, va_list ap) {
+        char errstr[512];
+        char extra[128];
+        size_t of, ofe;
+        rd_bool_t identical, suppress;
+        int state_duration_ms = (int)((rd_clock() - rkb->rkb_ts_state)/1000);
+
+
+        of = (size_t)rd_vsnprintf(errstr, sizeof(errstr), fmt, ap);
+        if (of > sizeof(errstr))
+                of = sizeof(errstr);
+
+        /* Provide more meaningful error messages in certain cases */
+        if (err == RD_KAFKA_RESP_ERR__TRANSPORT &&
+            !strcmp(errstr, "Disconnected")) {
+                if (rkb->rkb_state == RD_KAFKA_BROKER_STATE_APIVERSION_QUERY) {
+                        /* A disconnect while requesting ApiVersion typically
+                         * means we're connecting to a SSL-listener as
+                         * PLAINTEXT, but may also be caused by connecting to
+                         * a broker that does not support ApiVersion (<0.10). */
+
+                        if (rkb->rkb_proto != RD_KAFKA_PROTO_SSL &&
+                            rkb->rkb_proto != RD_KAFKA_PROTO_SASL_SSL)
+                                rd_kafka_broker_set_error(
+                                        rkb, level, err,
+                                        "Disconnected while requesting "
+                                        "ApiVersion: "
+                                        "might be caused by incorrect "
+                                        "security.protocol configuration or "
+                                        "broker version is < 0.10 "
+                                        "(see api.version.request)",
+                                        ap/*ignored*/);
+                        else
+                                rd_kafka_broker_set_error(
+                                        rkb, level, err,
+                                        "Disconnected while requesting "
+                                        "ApiVersion: "
+                                        "might be caused by broker version "
+                                        "< 0.10 (see api.version.request)",
+                                        ap/*ignored*/);
+                        return;
+
+                } else if (rkb->rkb_state == RD_KAFKA_BROKER_STATE_UP &&
+                           state_duration_ms < 2000/*2s*/ &&
+                           rkb->rkb_rk->rk_conf.security_protocol !=
+                           RD_KAFKA_PROTO_SASL_SSL &&
+                           rkb->rkb_rk->rk_conf.security_protocol !=
+                           RD_KAFKA_PROTO_SASL_PLAINTEXT) {
+                        /* If disconnected shortly after transitioning to UP
+                         * state it typically means the broker listener is
+                         * configured for SASL authentication but the client
+                         * is not. */
+                        rd_kafka_broker_set_error(
```

Comment:
👍 

The broker might require SASL authentication or it may be misconfgured.

Reply:
I think that suggesting that the broker is misconfigured is outside the scope of the client.